ServerExceptionMapper not called

### Describe the bug

ServerExceptionMappers are supposed to be always consulted if they perfectly match the exception.
But if proactive authentication is turned on, they are not.

### Expected behavior

Customt ExceptionMapper should be called if authentication fails

### Actual behavior

they are not.

### How to Reproduce?

run this [test](https://github.com/mschirmacher/QuarkusAuthorizationPolicySuspendBug/blob/noExceptionMapperCalled/src/test/kotlin/org/acme/GreetingResourceTest.kt)
If the custom exception mapper is called, the test should be green.
Note the config on line 29 of the test - if set to false the test is green!

Architectural overview:
The test sets up a wiremock response, that returns invalid json.
There is a [custom identity provider](https://github.com/mschirmacher/QuarkusAuthorizationPolicySuspendBug/blob/noExceptionMapperCalled/src/main/kotlin/org/acme/security/auth/CustomIdentityProvider.kt), that will call an external service (the one mocked by wiremock) for authentication.
As the response cannot be deserialized, an exception is thrown which SHOULD be handled by the matching [ServerExeptionMapper](https://github.com/mschirmacher/QuarkusAuthorizationPolicySuspendBug/blob/noExceptionMapperCalled/src/main/kotlin/org/acme/exception/ServerExceptionMapper.kt) - but its not.



### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

(my unprofessional debugging attemts reveald, that [this]{io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder} class is the bad guy as it doesnt seem to be prepared to call custom exception mappers)

When I run the test I get:

```posh
java.lang.AssertionError: 1 expectation failed.
Expected status code <234> but was <500>.

	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:73)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:108)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:57)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure.validate(ResponseSpecificationImpl.groovy:512)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure$validate$1.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
	at io.restassured.internal.ResponseSpecificationImpl.validateResponseIfRequired(ResponseSpecificationImpl.groovy:696)
	at io.restassured.internal.ResponseSpecificationImpl.this$2$validateResponseIfRequired(ResponseSpecificationImpl.groovy)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:198)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:62)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:135)
	at io.restassured.specification.ResponseSpecification$statusCode$0.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:143)
	at io.restassured.internal.ValidatableResponseOptionsImpl.statusCode(ValidatableResponseOptionsImpl.java:89)
	at org.acme.GreetingResourceTest.testHelloEndpoint(GreetingResourceTest.kt:53)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:967)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:817)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	Suppressed: org.jboss.resteasy.reactive.ClientWebApplicationException: Received: 'HTTP 200 OK' when invoking REST Client method: 'org.acme.security.auth.ExternalServiceClient#getTokenInfo'
		at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.unwrapException(RestClientRequestContext.java:205)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.handleException(AbstractResteasyReactiveContext.java:331)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:175)
		at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext$1.lambda$execute$0(RestClientRequestContext.java:324)
		at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
		at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
		at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:50)
		at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
		at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
		at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:1583)
	Caused by: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'invalid': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 9]
		at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2584)
		at com.fasterxml.jackson.core.JsonParser._constructReadException(JsonParser.java:2610)
		at com.fasterxml.jackson.core.JsonParser._constructReadException(JsonParser.java:2618)
		at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:825)
		at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._reportInvalidToken(UTF8StreamJsonParser.java:3662)
		at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._handleUnexpectedValue(UTF8StreamJsonParser.java:2749)
		at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._nextTokenNotInObject(UTF8StreamJsonParser.java:867)
		at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:753)
		at com.fasterxml.jackson.databind.ObjectReader._initForReading(ObjectReader.java:357)
		at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2115)
		at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1501)
		at io.quarkus.rest.client.reactive.jackson.runtime.serialisers.ClientJacksonMessageBodyReader.readFrom(ClientJacksonMessageBodyReader.java:51)
		at org.jboss.resteasy.reactive.client.impl.ClientReaderInterceptorContextImpl.proceed(ClientReaderInterceptorContextImpl.java:86)
		at org.jboss.resteasy.reactive.client.impl.ClientSerialisers.invokeClientReader(ClientSerialisers.java:160)
		at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.readEntity(RestClientRequestContext.java:219)
		at org.jboss.resteasy.reactive.client.handlers.ClientResponseCompleteRestHandler.mapToResponse(ClientResponseCompleteRestHandler.java:113)
		at org.jboss.resteasy.reactive.client.handlers.ClientResponseCompleteRestHandler.handle(ClientResponseCompleteRestHandler.java:35)
		at org.jboss.resteasy.reactive.client.handlers.ClientResponseCompleteRestHandler.handle(ClientResponseCompleteRestHandler.java:31)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.invokeHandler(AbstractResteasyReactiveContext.java:231)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
		... 12 more

```


which tells me that:

```kotlin
@RegisterRestClient(configKey = "foo")
interface ExternalServiceClient {

    @GET
    @Path("/{token}")
    fun getTokenInfo(
        @PathParam("token") token: String
    ): Uni<TokenInfo>
}

data class TokenInfo(val isValid: Boolean)

```

is not properly implemented.

As this is an exception originating from the REST Client, `@ServerExceptionMapper` is not part of picture

--- a/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/AbstractCustomExceptionMapperTest.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/AbstractCustomExceptionMapperTest.java
@@ -1,0 +1,164 @@
+package io.quarkus.resteasy.test.security;
+
+import java.util.Map;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.core.SecurityContext;
+import jakarta.ws.rs.ext.ExceptionMapper;
+import jakarta.ws.rs.ext.Provider;
+
+import org.hamcrest.Matchers;
+import org.jboss.resteasy.spi.UnhandledException;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.security.identity.AuthenticationRequestContext;
+import io.quarkus.security.identity.IdentityProvider;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.identity.SecurityIdentityAugmentor;
+import io.quarkus.security.identity.request.UsernamePasswordAuthenticationRequest;
+import io.quarkus.security.runtime.QuarkusPrincipal;
+import io.quarkus.security.runtime.QuarkusSecurityIdentity;
+import io.quarkus.vertx.http.runtime.security.HttpSecurityUtils;
+import io.restassured.RestAssured;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Tests internal server errors and other custom exceptions raised during
+ * proactive authentication can be handled by the exception mappers.
+ * For lazy authentication, it is important that these exceptions raised during authentication
+ * required by HTTP permissions are also propagated.
+ */
+public abstract class AbstractCustomExceptionMapperTest {
+
+    @Test
+    public void testNoExceptions() {
+        RestAssured.given()
+                .auth().preemptive().basic("gaston", "gaston-password")
+                .get("/hello")
+                .then()
+                .statusCode(200)
+                .body(Matchers.is("Hello Gaston"));
+        RestAssured.given()
+                .get("/hello")
+                .then()
+                .statusCode(401);
+    }
+
+    @Test
+    public void testUnhandledRuntimeException() {
+        // UnhandledRuntimeException has no exception mapper therefore RESTEasy would wrap it in UnhandledException
+        // if we started RESTEasy even though there is no matching exception mapper
+        RestAssured.given()
+                .auth().preemptive().basic("gaston", "gaston-password")
+                .header("fail-unhandled", "true")
+                .get("/hello")
+                .then()
+                .statusCode(500)
+                .body(Matchers.not(Matchers.is(UnhandledException.class.getName())))
+                .body(Matchers.containsString(UnhandledRuntimeException.class.getName()))
+                .body(Matchers.containsString("Expected unhandled failure"));
+    }
+
+    @Test
+    public void testCustomExceptionInIdentityProvider() {
+        RestAssured.given()
+                .auth().preemptive().basic("gaston", "gaston-password")
+                .header("fail-authentication", "true")
+                .get("/hello")
+                .then()
+                .statusCode(500)
+                .body(Matchers.is("Expected authentication failure"));
+    }
+
+    @Test
+    public void testCustomExceptionInIdentityAugmentor() {
+        RestAssured.given()
+                .auth().preemptive().basic("gaston", "gaston-password")
+                .header("fail-augmentation", "true")
+                .get("/hello")
+                .then()
+                .statusCode(500)
+                .body(Matchers.is("Expected identity augmentation failure"));
+    }
+
+    @Path("/hello")
+    public static class HelloResource {
+        @GET
+        public String hello(@Context SecurityContext context) {
+            var principalName = context.getUserPrincipal() == null ? "" : " " + context.getUserPrincipal().getName();
+            return "Hello" + principalName;
+        }
+
+    }
+
+    @Provider
+    public static class CustomRuntimeExceptionMapper implements ExceptionMapper<CustomRuntimeException> {
+        @Override
+        public Response toResponse(CustomRuntimeException exception) {
+            return Response.serverError().entity(exception.getMessage()).build();
+        }
+    }
+
+    @ApplicationScoped
+    public static class CustomIdentityAugmentor implements SecurityIdentityAugmentor {
+        @Override
+        public Uni<SecurityIdentity> augment(SecurityIdentity securityIdentity,
+                AuthenticationRequestContext authenticationRequestContext) {
+            return augment(securityIdentity, authenticationRequestContext, Map.of());
+        }
+
+        @Override
+        public Uni<SecurityIdentity> augment(SecurityIdentity identity, AuthenticationRequestContext context,
+                Map<String, Object> attributes) {
+            final RoutingContext routingContext = HttpSecurityUtils.getRoutingContextAttribute(attributes);
+            if (routingContext.request().headers().contains("fail-augmentation")) {
+                return Uni.createFrom().failure(new CustomRuntimeException("Expected identity augmentation failure"));
+            }
+            return Uni.createFrom().item(identity);
+        }
+    }
+
+    public static class CustomRuntimeException extends RuntimeException {
+        public CustomRuntimeException(String message) {
+            super(message);
+        }
+    }
+
+    public static class UnhandledRuntimeException extends RuntimeException {
+        public UnhandledRuntimeException(String message) {
+            super(message);
+        }
+    }
+
+    @ApplicationScoped
+    public static class BasicIdentityProvider implements IdentityProvider<UsernamePasswordAuthenticationRequest> {
+
+        @Override
+        public Class<UsernamePasswordAuthenticationRequest> getRequestType() {
+            return UsernamePasswordAuthenticationRequest.class;
+        }
+
+        @Override
+        public Uni<SecurityIdentity> authenticate(UsernamePasswordAuthenticationRequest authRequest,
+                AuthenticationRequestContext authRequestCtx) {
+            if (!"gaston".equals(authRequest.getUsername())) {
+                return Uni.createFrom().nullItem();
+            }
+
+            final RoutingContext routingContext = HttpSecurityUtils.getRoutingContextAttribute(authRequest);
+            if (routingContext.request().headers().contains("fail-authentication")) {
+                return Uni.createFrom().failure(new CustomRuntimeException("Expected authentication failure"));
+            }
+            if (routingContext.request().headers().contains("fail-unhandled")) {
+                return Uni.createFrom().failure(new UnhandledRuntimeException("Expected unhandled failure"));
+            }
+            return Uni.createFrom()
+                    .item(QuarkusSecurityIdentity.builder().setPrincipal(new QuarkusPrincipal("Gaston")).build());
+        }
+    }
+}

--- a/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/LazyAuthCustomExceptionMapperTest.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/LazyAuthCustomExceptionMapperTest.java
@@ -1,0 +1,18 @@
+package io.quarkus.resteasy.test.security;
+
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class LazyAuthCustomExceptionMapperTest extends AbstractCustomExceptionMapperTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest().withApplicationRoot(jar -> jar
+            .addAsResource(new StringAsset("""
+                    quarkus.http.auth.permission.authentication.paths=*
+                    quarkus.http.auth.permission.authentication.policy=authenticated
+                    quarkus.http.auth.proactive=false
+                    """), "application.properties"));
+
+}

--- a/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/ProactiveAuthCustomExceptionMapperTest.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/ProactiveAuthCustomExceptionMapperTest.java
@@ -1,0 +1,17 @@
+package io.quarkus.resteasy.test.security;
+
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ProactiveAuthCustomExceptionMapperTest extends AbstractCustomExceptionMapperTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest().withApplicationRoot(jar -> jar
+            .addAsResource(new StringAsset("""
+                    quarkus.http.auth.permission.authentication.paths=*
+                    quarkus.http.auth.permission.authentication.policy=authenticated
+                    """), "application.properties"));
+
+}

--- a/extensions/resteasy-classic/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/standalone/ResteasyStandaloneRecorder.java
+++ b/extensions/resteasy-classic/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/standalone/ResteasyStandaloneRecorder.java
@@ -1,6 +1,9 @@
 package io.quarkus.resteasy.runtime.standalone;
 
 import static io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler.extractRootCause;
+import static io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler.isOtherAuthenticationFailure;
+import static io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler.markIfOtherAuthenticationFailure;
+import static io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler.removeMarkAsOtherAuthenticationFailure;
 
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Proxy;
@@ -37,6 +40,7 @@ import io.quarkus.security.AuthenticationException;
 import io.quarkus.security.AuthenticationFailedException;
 import io.quarkus.security.AuthenticationRedirectException;
 import io.quarkus.security.ForbiddenException;
+import io.quarkus.security.UnauthorizedException;
 import io.quarkus.vertx.http.runtime.HttpBuildTimeConfig;
 import io.quarkus.vertx.http.runtime.HttpCompressionHandler;
 import io.quarkus.vertx.http.runtime.HttpConfiguration;
@@ -163,8 +167,16 @@ public class ResteasyStandaloneRecorder {
                         }
                     }
 
-                    if (request.failure() instanceof AuthenticationException
-                            || request.failure() instanceof ForbiddenException) {
+                    final Throwable failure = request.failure();
+                    final boolean isOtherAuthFailure = isOtherAuthenticationFailure(request)
+                            && isFailureHandledByExceptionMappers(failure);
+                    if (isOtherAuthFailure) {
+                        // prevent circular reference for unhandled exceptions
+                        // (which is unnecessary if everything here is done right)
+                        removeMarkAsOtherAuthenticationFailure(request);
+                        super.handle(request);
+                    } else if (failure instanceof AuthenticationException || failure instanceof UnauthorizedException
+                            || failure instanceof ForbiddenException) {
                         super.handle(request);
                     } else {
                         request.next();
@@ -177,6 +189,11 @@ public class ResteasyStandaloneRecorder {
                 }
             };
         }
+    }
+
+    private boolean isFailureHandledByExceptionMappers(Throwable failure) {
+        return failure != null && deployment != null
+                && deployment.getProviderFactory().getExceptionMapper(failure.getClass()) != null;
     }
 
     public Handler<RoutingContext> defaultAuthFailureHandler() {
@@ -204,6 +221,7 @@ public class ResteasyStandaloneRecorder {
 
                         @Override
                         public void accept(RoutingContext event, Throwable throwable) {
+                            markIfOtherAuthenticationFailure(event, throwable);
                             if (!event.failed()) {
                                 event.fail(extractRootCause(throwable));
                             }

--- a/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/AbstractCustomExceptionMapperTest.java
+++ b/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/AbstractCustomExceptionMapperTest.java
@@ -1,0 +1,157 @@
+package io.quarkus.resteasy.reactive.server.test.security;
+
+import java.util.Map;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.core.SecurityContext;
+
+import org.hamcrest.Matchers;
+import org.jboss.resteasy.reactive.server.ServerExceptionMapper;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.security.identity.AuthenticationRequestContext;
+import io.quarkus.security.identity.IdentityProvider;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.identity.SecurityIdentityAugmentor;
+import io.quarkus.security.identity.request.UsernamePasswordAuthenticationRequest;
+import io.quarkus.security.runtime.QuarkusPrincipal;
+import io.quarkus.security.runtime.QuarkusSecurityIdentity;
+import io.quarkus.vertx.http.runtime.security.HttpSecurityUtils;
+import io.restassured.RestAssured;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Tests internal server errors and other custom exceptions raised during
+ * proactive authentication can be handled by the exception mappers.
+ * For lazy authentication, it is important that these exceptions raised during authentication
+ * required by HTTP permissions are also propagated.
+ */
+public abstract class AbstractCustomExceptionMapperTest {
+
+    @Test
+    public void testNoExceptions() {
+        RestAssured.given()
+                .auth().preemptive().basic("gaston", "gaston-password")
+                .get("/hello")
+                .then()
+                .statusCode(200)
+                .body(Matchers.is("Hello Gaston"));
+        RestAssured.given()
+                .get("/hello")
+                .then()
+                .statusCode(401);
+    }
+
+    @Test
+    public void testUnhandledRuntimeException() {
+        RestAssured.given()
+                .auth().preemptive().basic("gaston", "gaston-password")
+                .header("fail-unhandled", "true")
+                .get("/hello")
+                .then()
+                .statusCode(500)
+                .body(Matchers.containsString(UnhandledRuntimeException.class.getName()))
+                .body(Matchers.containsString("Expected unhandled failure"));
+    }
+
+    @Test
+    public void testCustomExceptionInIdentityProvider() {
+        RestAssured.given()
+                .auth().preemptive().basic("gaston", "gaston-password")
+                .header("fail-authentication", "true")
+                .get("/hello")
+                .then()
+                .statusCode(500)
+                .body(Matchers.is("Expected authentication failure"));
+    }
+
+    @Test
+    public void testCustomExceptionInIdentityAugmentor() {
+        RestAssured.given()
+                .auth().preemptive().basic("gaston", "gaston-password")
+                .header("fail-augmentation", "true")
+                .get("/hello")
+                .then()
+                .statusCode(500)
+                .body(Matchers.is("Expected identity augmentation failure"));
+    }
+
+    @Path("/hello")
+    public static class HelloResource {
+        @GET
+        public String hello(@Context SecurityContext context) {
+            var principalName = context.getUserPrincipal() == null ? "" : " " + context.getUserPrincipal().getName();
+            return "Hello" + principalName;
+        }
+    }
+
+    public static class Mappers {
+        @ServerExceptionMapper(CustomRuntimeException.class)
+        public Response toResponse(CustomRuntimeException exception) {
+            return Response.serverError().entity(exception.getMessage()).build();
+        }
+    }
+
+    @ApplicationScoped
+    public static class CustomIdentityAugmentor implements SecurityIdentityAugmentor {
+        @Override
+        public Uni<SecurityIdentity> augment(SecurityIdentity securityIdentity,
+                AuthenticationRequestContext authenticationRequestContext) {
+            return augment(securityIdentity, authenticationRequestContext, Map.of());
+        }
+
+        @Override
+        public Uni<SecurityIdentity> augment(SecurityIdentity identity, AuthenticationRequestContext context,
+                Map<String, Object> attributes) {
+            final RoutingContext routingContext = HttpSecurityUtils.getRoutingContextAttribute(attributes);
+            if (routingContext.request().headers().contains("fail-augmentation")) {
+                return Uni.createFrom().failure(new CustomRuntimeException("Expected identity augmentation failure"));
+            }
+            return Uni.createFrom().item(identity);
+        }
+    }
+
+    public static class CustomRuntimeException extends RuntimeException {
+        public CustomRuntimeException(String message) {
+            super(message);
+        }
+    }
+
+    public static class UnhandledRuntimeException extends RuntimeException {
+        public UnhandledRuntimeException(String message) {
+            super(message);
+        }
+    }
+
+    @ApplicationScoped
+    public static class BasicIdentityProvider implements IdentityProvider<UsernamePasswordAuthenticationRequest> {
+
+        @Override
+        public Class<UsernamePasswordAuthenticationRequest> getRequestType() {
+            return UsernamePasswordAuthenticationRequest.class;
+        }
+
+        @Override
+        public Uni<SecurityIdentity> authenticate(UsernamePasswordAuthenticationRequest authRequest,
+                AuthenticationRequestContext authRequestCtx) {
+            if (!"gaston".equals(authRequest.getUsername())) {
+                return Uni.createFrom().nullItem();
+            }
+
+            final RoutingContext routingContext = HttpSecurityUtils.getRoutingContextAttribute(authRequest);
+            if (routingContext.request().headers().contains("fail-authentication")) {
+                return Uni.createFrom().failure(new CustomRuntimeException("Expected authentication failure"));
+            }
+            if (routingContext.request().headers().contains("fail-unhandled")) {
+                return Uni.createFrom().failure(new UnhandledRuntimeException("Expected unhandled failure"));
+            }
+            return Uni.createFrom()
+                    .item(QuarkusSecurityIdentity.builder().setPrincipal(new QuarkusPrincipal("Gaston")).build());
+        }
+    }
+}

--- a/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/LazyAuthCustomExceptionMapperTest.java
+++ b/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/LazyAuthCustomExceptionMapperTest.java
@@ -1,0 +1,18 @@
+package io.quarkus.resteasy.reactive.server.test.security;
+
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class LazyAuthCustomExceptionMapperTest extends AbstractCustomExceptionMapperTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest().withApplicationRoot(jar -> jar
+            .addAsResource(new StringAsset("""
+                    quarkus.http.auth.permission.authentication.paths=*
+                    quarkus.http.auth.permission.authentication.policy=authenticated
+                    quarkus.http.auth.proactive=false
+                    """), "application.properties"));
+
+}

--- a/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/ProactiveAuthCustomExceptionMapperTest.java
+++ b/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/ProactiveAuthCustomExceptionMapperTest.java
@@ -1,0 +1,17 @@
+package io.quarkus.resteasy.reactive.server.test.security;
+
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ProactiveAuthCustomExceptionMapperTest extends AbstractCustomExceptionMapperTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest().withApplicationRoot(jar -> jar
+            .addAsResource(new StringAsset("""
+                    quarkus.http.auth.permission.authentication.paths=*
+                    quarkus.http.auth.permission.authentication.policy=authenticated
+                    """), "application.properties"));
+
+}

--- a/extensions/resteasy-reactive/rest/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java
+++ b/extensions/resteasy-reactive/rest/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java
@@ -1,6 +1,9 @@
 package io.quarkus.resteasy.reactive.server.runtime;
 
 import static io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler.extractRootCause;
+import static io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler.isOtherAuthenticationFailure;
+import static io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler.markIfOtherAuthenticationFailure;
+import static io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler.removeMarkAsOtherAuthenticationFailure;
 
 import java.io.Closeable;
 import java.lang.reflect.InvocationTargetException;
@@ -66,6 +69,7 @@ import io.quarkus.security.AuthenticationException;
 import io.quarkus.security.AuthenticationFailedException;
 import io.quarkus.security.AuthenticationRedirectException;
 import io.quarkus.security.ForbiddenException;
+import io.quarkus.security.UnauthorizedException;
 import io.quarkus.security.identity.CurrentIdentityAssociation;
 import io.quarkus.vertx.http.runtime.CurrentVertxRequest;
 import io.quarkus.vertx.http.runtime.HttpBuildTimeConfig;
@@ -231,14 +235,25 @@ public class ResteasyReactiveRecorder extends ResteasyReactiveCommonRecorder imp
                     }
                 }
 
-                if (event.failure() instanceof AuthenticationException
-                        || event.failure() instanceof ForbiddenException) {
-                    restInitialHandler.beginProcessing(event, event.failure());
+                final Throwable failure = event.failure();
+                final boolean isOtherAuthFailure = isOtherAuthenticationFailure(event)
+                        && isFailureHandledByExceptionMappers(failure);
+                if (isOtherAuthFailure) {
+                    removeMarkAsOtherAuthenticationFailure(event);
+                    restInitialHandler.beginProcessing(event, failure);
+                } else if (failure instanceof AuthenticationException
+                        || failure instanceof UnauthorizedException || failure instanceof ForbiddenException) {
+                    restInitialHandler.beginProcessing(event, failure);
                 } else {
                     event.next();
                 }
             }
         };
+    }
+
+    private boolean isFailureHandledByExceptionMappers(Throwable throwable) {
+        return currentDeployment != null
+                && currentDeployment.getExceptionMapper().getExceptionMapper(throwable.getClass(), null, null) != null;
     }
 
     /**
@@ -420,6 +435,7 @@ public class ResteasyReactiveRecorder extends ResteasyReactiveCommonRecorder imp
 
         @Override
         public void accept(RoutingContext event, Throwable throwable) {
+            markIfOtherAuthenticationFailure(event, throwable);
             if (!event.failed()) {
                 event.fail(extractRootCause(throwable));
             }

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java
@@ -33,6 +33,7 @@ import io.quarkus.runtime.RuntimeValue;
 import io.quarkus.runtime.annotations.Recorder;
 import io.quarkus.runtime.configuration.ConfigurationException;
 import io.quarkus.security.AuthenticationCompletionException;
+import io.quarkus.security.AuthenticationException;
 import io.quarkus.security.AuthenticationFailedException;
 import io.quarkus.security.AuthenticationRedirectException;
 import io.quarkus.security.identity.SecurityIdentity;
@@ -171,6 +172,12 @@ public class HttpSecurityRecorder {
 
     public static abstract class DefaultAuthFailureHandler implements BiConsumer<RoutingContext, Throwable> {
 
+        /**
+         * A {@link RoutingContext#get(String)} key added for exceptions raised during authentication that are not
+         * the {@link io.quarkus.security.AuthenticationException}.
+         */
+        private static final String OTHER_AUTHENTICATION_FAILURE = "io.quarkus.vertx.http.runtime.security.other-auth-failure";
+
         protected DefaultAuthFailureHandler() {
         }
 
@@ -206,6 +213,7 @@ public class HttpSecurityRecorder {
                 event.response().headers().set("Pragma", "no-cache");
                 proceed(throwable);
             } else {
+                event.put(OTHER_AUTHENTICATION_FAILURE, Boolean.TRUE);
                 event.fail(throwable);
             }
         }
@@ -226,6 +234,20 @@ public class HttpSecurityRecorder {
                 }
             }
             return throwable;
+        }
+
+        public static void markIfOtherAuthenticationFailure(RoutingContext event, Throwable throwable) {
+            if (!(throwable instanceof AuthenticationException)) {
+                event.put(OTHER_AUTHENTICATION_FAILURE, Boolean.TRUE);
+            }
+        }
+
+        public static void removeMarkAsOtherAuthenticationFailure(RoutingContext event) {
+            event.remove(OTHER_AUTHENTICATION_FAILURE);
+        }
+
+        public static boolean isOtherAuthenticationFailure(RoutingContext event) {
+            return Boolean.TRUE.equals(event.get(OTHER_AUTHENTICATION_FAILURE));
         }
     }
 

